subcommand/connect-init/command_test.go

```diff
@@ -40,18 +61,249 @@ func TestRun_FlagValidation(t *testing.T) {
+	time.Sleep(time.Second * 1)
+	// Register Consul services.
+	testConsulServices := []api.AgentServiceRegistration{consulCountingSvc, consulCountingSvcSidecar}
+	for _, svc := range testConsulServices {
+		require.NoError(t, consulClient.Agent().ServiceRegister(&svc))
+		time.Sleep(time.Second * 2)
+	}
+
+	// Assert that it exits cleanly or timeout.
+	select {
+	case exitCode := <-exitChan:
+		require.Equal(t, 0, exitCode)
+	case <-time.After(time.Second * 10):
+		// Fail if the stopCh was not caught.
+		require.Fail(t, "timeout waiting for command to exit")
+	}
+	// Validate that we hit the retry logic when proxy service is not registered yet.
+	require.Contains(t, ui.OutputWriter.String(), "Unable to find registered services; Retrying")
+
+	// Validate contents of proxyFile.
+	data, err := ioutil.ReadFile(proxyFile)
+	require.NoError(t, err)
+	require.Contains(t, string(data), "counting-counting-sidecar-proxy")
+}
+
+// TestRun_invalidProxyFile validates that we correctly fail in case the proxyid file
+// is not writable. This functions as coverage for both ACL and non-ACL codepaths.
+func TestRun_invalidProxyFile(t *testing.T) {
+	t.Parallel()
+	// This is the output file for the proxyid.
+	randFileName := fmt.Sprintf("/foo/%d/%d", rand.Int(), rand.Int())
+
+	// Start Consul server.
+	server, err := testutil.NewTestServerConfigT(t, nil)
+	defer server.Stop()
+	require.NoError(t, err)
+	server.WaitForLeader(t)
+	consulClient, err := api.NewClient(&api.Config{Address: server.HTTPAddr})
+	require.NoError(t, err)
+
+	// Register Consul services.
+	testConsulServices := []api.AgentServiceRegistration{consulCountingSvc, consulCountingSvcSidecar}
+	for _, svc := range testConsulServices {
+		require.NoError(t, consulClient.Agent().ServiceRegister(&svc))
+	}
+	ui := cli.NewMockUi()
+	cmd := Command{
+		UI:                                 ui,
+		consulClient:                       consulClient,
+		ProxyIDFile:                        randFileName,
+		ServiceRegistrationPollingAttempts: 3,
+	}
+	expErr := fmt.Sprintf("Unable to write proxyid out: open %s: no such file or directory\n", randFileName)
+	code := cmd.Run(defaultTestFlags)
+	require.Equal(t, 1, code)
+	require.Equal(t, expErr, ui.ErrorWriter.String())
+}
+
+// TestRun_ServiceRegistrationFailsWithBadServerResponses tests error handling with invalid server responses.
+func TestRun_ServiceRegistrationFailsWithBadServerResponses(t *testing.T) {
+	cases := []struct {
+		name                    string
+		loginResponse           string
+		getServicesListResponse string
+		expErr                  string
+	}{
+		{
+			name:          "acls enabled, acl login response invalid",
+			loginResponse: "",
+			expErr:        "Hit maximum retries for consul login",
+		},
+		{
+			name:                    "acls enabled, get service response invalid",
+			loginResponse:           testLoginResponse,
+			getServicesListResponse: "",
+			expErr:                  "Timed out waiting for service registration",
+		},
```

Comment:
With ACLs enabled if loginResponse is "" then it would fail in login and return failed, it would never issue a request for the getServiceListResponse. So that becomes the first test. I could add a valid response for getServicesListResponse to the first test ?